druid/src/widget/focus_scope.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 The druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A focus scope widget.
+
+use druid::widget::prelude::*;
+use druid::{Data, FocusScopeNode, Point, Rect, Widget, WidgetPod};
+
+/// A Widget that serves as a scope for its descendants,
+/// restricting focus traversal to the scoped controls.
+pub struct FocusScope<T> {
+    child: WidgetPod<T, Box<dyn Widget<T>>>,
+    focus_scope_node: FocusScopeNode,
+}
+
+impl<T: Data> FocusScope<T> {
+    /// Create a new FocusScope widget with a child
+    pub fn new(child: impl Widget<T> + 'static) -> Self {
+        FocusScope {
+            child: WidgetPod::new(child).boxed(),
+            focus_scope_node: FocusScopeNode { widget_id: None },
+        }
+    }
+}
+
+impl<T: Data> Widget<T> for FocusScope<T> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        let previous_focus_scope = ctx.focus_scope();
+        ctx.set_focus_scope_node(self.focus_scope_node);
```

Comment:
Just an observation: an idiomatic way to get around writing these dual methods is to have a method on the `ctx`, like `with_focus_scope`, that takes a closure, similar to how `PaintCtx::with_save` works; then you could write,

```rust
ctx.with_focus_scope(self.focus_scope_node, |ctx| {
    self.child_event(ctx, event, data, env)
});

Reply:
Oh yeah, thanks a lot! I actually don't like this solution with dual methods, but my attempt using `Drop` was failed because of the borrow checker, and I haven't come across any different solution. I'll fix this.